### DISCUSSION TITLE
Filament FileUpload with Tenancy for Laravel

### DISCUSSION THREAD

**Discussion by Satal** [filamentphp/filament#12415] [Help]
### Package

Panel builder

### Package Version

v3.1.47

### How can we help you?

I have an application that I'm developing that uses Filament and Tenancy for Laravel (stancl/tenancy package v3.8.2, https://tenancyforlaravel.com/). 

In my project, I have a `TenantProjectResource`
```
public static function form(Form $form): Form
    {
        return $form
            ->schema([
                Forms\Components\TextInput::make('name')
                    ->required(),
                Forms\Components\FileUpload::make('image')
                    ->image()
                    ->label('Project Default Image')
                    ->disk('public')
                    ->columns(1)
                    ->directory('projects')
                    ->moveFiles()
                    ->maxSize(2048)
                    ->columnSpanFull(),
            ]);
    }
```

However, when I go to upload an image I get an error `Unable to retrieve the file_size for file at location: livewire-tmp/3CFKrO7tvOwRjB67kEKKS99122XY9C-metaZ3lwdS5jb20uanBn-.jpg.`
However, I can find the file at `storage/app/public/livewire-tmp/3CFKrO7tvOwRjB67kEKKS99122XY9C-metaZ3lwdS5jb20uanBn-.jpg` 

The `config\filesystems.php` disks section includes:
```
        'public' => [
            'driver' => 'local',
            'root' => storage_path('app/public'),
            'url' => env('APP_URL').'/storage',
            'visibility' => 'public',
            'throw' => false,
        ],
```

I've tried just removing the `maxSize` and get another error, so I'm guessing it's not just an issue with the `maxSize` and a deeper issue, but I'm running out of ideas on how to narrow this down.

I would greatly appreciate any advice anyone can give me.

**Comment by leandrocfe** [comment]
Publish the livewire config: `php artisan livewire:publish --config`

Try this

```php
'temporary_file_upload' => [
    // ...
    'directory' => 'public/livewire-tmp/',
],
```

**Reply by Satal** [reply]
Thanks @leandrocfe.
I've updated my `config/livewire.php` to:
```php
        'temporary_file_upload' => [
            'disk' => 'public',
            // Example: 'local', 's3'              | Default: 'default'
            'rules' => null,
            // Example: ['file', 'mimes:png,jpg']  | Default: ['required', 'file', 'max:12288'] (12MB)
            'directory' => 'public/livewire-tmp/',
            // Example: 'tmp'                      | Default: 'livewire-tmp'
            'middleware' => ['throttle:60,1', 'universal', InitializeTenancyByDomain::class],
            // Example: 'throttle:5,1'             | Default: 'throttle:60,1'
            'preview_mimes' => [   // Supported file types for temporary pre-signed file URLs...
                'png', 'gif', 'bmp', 'svg', 'wav', 'mp4',
                'mov', 'avi', 'wmv', 'mp3', 'm4a', 'jpg',
                'jpeg', 'mpga', 'webp', 'wma',
            ],
            'max_upload_time' => 5,
            // Max duration (in minutes) before an upload is invalidated...
        ],
```

Unfortunately, it still doesn't work 😢 I've also tried changing it to `storage/app/public/livewire-tmp` and `app/public/livewire-tmp` in case that worked but it hasn't unfortunately.

**Reply by leandrocfe** [reply]
Did you try it without `->moveFiles()`?

**Reply by Satal** [reply]
I've just tried it, however, I get the same error. I'm guessing that this is because of the behaviour I'm seeing in https://github.com/filamentphp/filament/discussions/12415#discussioncomment-9178111

**Comment by Satal** [comment]
So as an update, I've gone into the `LocalFilesystemAdapter` to try and debug, changing the `fileSize` function to:
```php
    public function fileSize(string $path): FileAttributes
    {
        $location = $this->prefixer->prefixPath($path);
        error_clear_last();
dd($location, is_file($location)); // new line to test
        if (is_file($location) && ($fileSize = @filesize($location)) !== false) {
            return new FileAttributes($path, $fileSize);
        }

        throw UnableToRetrieveMetadata::fileSize($path, error_get_last()['message'] ?? '');
    }
```

The output of that when I upload a file is:
```bash
"/var/www/html/storage/tenant1/app/public/livewire-tmp/cKQr3ZFo3nhUl6aaKaR5IqfBAhJKzw-metaYWZsb2dvLnBuZw==-.png"// vendor/league/flysystem-local/LocalFilesystemAdapter.php:440
false // vendor/league/flysystem-local/LocalFilesystemAdapter.php:440
```
(`config/livewire.php` `temporary_file_upload.directory` is 'livewire-tmp/')

So it looks like the file is uploaded, it is being uploaded to `storage/app/[temporary_file_upload.directory]`, however, when it passes the file to `fileSize` it is looking in `storage/tenant[id]/app/public/[temporary_file_upload.directory]`.

I'm guessing there's a chance I've missed somewhere that has files uploaded to a tenant's temporary directory.

**Reply by Mahdi-Sahib** [reply]
I have exactly the same issue. Is there a solution?

**Reply by abdallah84** [reply]
Hi @Satal 
do you solve the issue ?

**Comment by Mohammed-800** [comment]
@Satal How to use private disk in Tenancy for laravel?

**Reply by long-blade** [reply]
i am also interested in that! Any solutions ?

**Comment by mirahy** [comment]
https://tenancyforlaravel.com/docs/v3/integrations/livewire/
Tutorial:
https://laracasts.com/discuss/channels/livewire/misplacement-of-laravel-livewire-tenancy-file-uploads-into-central-storage-instead-of-tenant-specific-storage
